webrtc-c/canary/src/Peer.cpp

```diff
@@ -533,13 +555,145 @@ STATUS Peer::writeFrame(PFrame pFrame, MEDIA_STREAM_TRACK_KIND kind)
     STATUS retStatus = STATUS_SUCCESS;
 
     auto& transceivers = kind == MEDIA_STREAM_TRACK_KIND_VIDEO ? this->videoTransceivers : this->audioTransceivers;
-
+    if (kind == MEDIA_STREAM_TRACK_KIND_VIDEO) {
+        std::lock_guard<std::mutex> lock(this->countUpdateMutex);
+        if (this->recorded.load()) {
+            this->canaryOutgoingRTPMetricsContext.videoFramesGenerated = 0;
+            this->canaryOutgoingRTPMetricsContext.videoBytesGenerated = 0;
+            this->recorded = FALSE;
+        }
+        this->canaryOutgoingRTPMetricsContext.videoFramesGenerated++;
+        this->canaryOutgoingRTPMetricsContext.videoBytesGenerated += pFrame->size;
+    }
     for (auto& transceiver : transceivers) {
-        CHK_LOG_ERR(::writeFrame(transceiver, pFrame));
+        retStatus = ::writeFrame(transceiver, pFrame);
+        if (retStatus == STATUS_SRTP_NOT_READY_YET || retStatus == STATUS_SUCCESS) {
+            // do nothing
+        }
+    }
+
+CleanUp:
+
+    return retStatus;
+}
+
+VOID Peer::setStatsType(RTC_STATS_TYPE statsType)
+{
+    this->canaryMetrics.requestedTypeOfStats = statsType;
+}
+
+RTC_STATS_TYPE Peer::getStatsType()
+{
+    return this->canaryMetrics.requestedTypeOfStats;
+}
+
+STATUS Peer::populateOutgoingRtpMetricsContext()
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    DOUBLE currentDuration = 0;
+    DOUBLE videoFramesGenerated = 0;
+    DOUBLE videoBytesGenerated = 0;
+
+    {
+        std::lock_guard<std::mutex> lock(this->countUpdateMutex);
+        videoFramesGenerated = (DOUBLE) this->canaryOutgoingRTPMetricsContext.videoFramesGenerated;
+        videoBytesGenerated = (DOUBLE) this->canaryOutgoingRTPMetricsContext.videoBytesGenerated;
     }
 
+    currentDuration = (DOUBLE)(this->canaryMetrics.timestamp - this->canaryOutgoingRTPMetricsContext.prevTs) / HUNDREDS_OF_NANOS_IN_A_SECOND;
```

Comment:
Important: since this block is not guarded with a `countUpdateMutex` lock, it's possible that we'll lose some data for `videoFramesGenerated` and `videoBytesGenerated`. Between 603 and 616, it's possible that there's a new call to `writeFrame` and increment these variables.

Reply:
Agreed. Removed `videoFramesGenerated` and `videoBytesGenerated` and I am guarding the lines with the mutex.